File: course/section_4_stack/code/15-evaluate-reverse-polish-notation.py
from typing import List
from math import floor, ceil


class Solution:
    def round_to_zero(self, n: float) -> int:
        if n >= 0:
            return floor(n)
        return ceil(n)

    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if token not in ["+", "-", "*", "/"]:
                stack.append(int(token))
            else:
                num2 = int(stack.pop())
                num1 = int(stack.pop())
                if token == "+":
                    stack.append(num1 + num2)
                elif token == "-":
                    stack.append(num1 - num2)
                elif token == "*":
                    stack.append(num1 * num2)
                elif token == "/":
                    stack.append(self.round_to_zero(num1 / num2))
        return stack[0]

File: course/section_4_stack/code/15-evaluate-reverse-polish-notation.php
<?php

class Solution {

    /**
     * @param String[] $tokens
     * @return Integer
     */
    function evalRPN($tokens) {
        $stack = [];
        foreach ($tokens as $token) {
            if ($token === '+') {
                $stack[] = array_pop($stack) + array_pop($stack);
            } elseif ($token === '*') {
                $stack[] = array_pop($stack) * array_pop($stack);
            } elseif ($token === '/') {
                $num2 = array_pop($stack);
                $num1 = array_pop($stack);
                $stack[] = (int)($num1 / $num2);
            } elseif ($token === '-') {
                $num2 = array_pop($stack);
                $num1 = array_pop($stack);
                $stack[] =  $num1 - $num2;
            } else {
                $stack[] = (int)$token;
            }
        }
        return $stack[0];
    }
}
